### PR TITLE
Promote cpp/unsigned-difference-expression-compared-zero out of experimental

### PR DESCRIPTION
**Pull request by MathiasVP**
This is the final step of https://github.com/github/codeql-c-analysis-team/issues/210. I didn't promote the query to high precision yet as that's been moved to https://github.com/github/codeql-c-analysis-team/issues/226.

### CODE REVIEW

File: cpp/change-notes/2020-02-04-unsigned-difference-expression-compared-zero.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+lgtm
```
Comment (hubwriter) [inline cpp/change-notes/2020-02-04-unsigned-difference-expression-compared-zero.md:1]: The other change notes all seem to be `lgtm,codescanning`.
No `codescanning` for this?

File: cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+This rule finds relational comparisons between the result of an unsigned subtraction and the value <code>0</code>.
+Such comparisons are likely wrong as the value of an unsigned subtraction can never be negative. So the
```
Comment (hubwriter) [inline cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp:8]:
```suggestion
Such comparisons are likely to be wrong as the value of an unsigned subtraction can never be negative. So the
```

File: cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+relational comparison ends up checking whether the result of the subtraction is equal to <code>0</code>.
+This is likely not what the programmer intended.
```
Comment (hubwriter) [inline cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp:10]:
```suggestion
This is probably not what the programmer intended.
```

File: cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+</p>
+
+</recommendation>
+<example>
+<sample src="UnsignedDifferenceExpressionComparedZero.c" />
+
+</example>
+<references>
+
+<li>CERT C Coding Standard:
```
Comment (hubwriter) [inline cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp:26]: Maybe this should be 
```suggestion
<li>SEI CERT C Coding Standard:
```
as per https://wiki.sei.cmu.edu/confluence/display/c

**Review by hubwriter — COMMENTED** [review]
LGTM - just a few minor observations.

File: cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp (hunk 4 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.qhelp:26]: Yes, I think so. Good catch! It seems we're not entirely consistent wrt. this. I'll fix the remaining ones in another PR.

File: cpp/change-notes/2020-02-04-unsigned-difference-expression-compared-zero.md (hunk 1 above)
Comment (MathiasVP) [reply cpp/change-notes/2020-02-04-unsigned-difference-expression-compared-zero.md:1]: No. As far as I know, code scanning only runs queries with @precision high, and this one is currently at medium. We hope that this query can be promoted to high precision in the future, but we want to try it out on LGTM first.

File: cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql [hunk 5]
```diff
@@ -16,16 +16,24 @@ import semmle.code.cpp.valuenumbering.GlobalValueNumbering
...
   or
   // The subtraction is guarded by a check of the form `left >= right`.
-  exists(GuardCondition guard, Expr left, Expr right |
-    left = globalValueNumber(sub.getLeftOperand()).getAnExpr() and
-    right = globalValueNumber(sub.getRightOperand()).getAnExpr() and
-    guard.controls(sub.getBasicBlock(), true) and
-    guard.ensuresLt(left, right, 0, sub.getBasicBlock(), false)
+  exists(GVN left, GVN right |
+    strictcount([left, globalValueNumber(sub.getLeftOperand())]) = 1 and
+    strictcount([right, globalValueNumber(sub.getRightOperand())]) = 1 and
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql:35]: What are these two lines doing?

Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql:35]: I should thank you for those, actually! I found it on [Slack](https://github.slack.com/archives/CPCFXL8P3/p1605720700336600) as a response from `@p0` to a question you asked on Slack on how to unbind values of database types.

Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql:35]: Cool!  Might be worth adding a comment stating that this is an unbind?

Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-191/UnsignedDifferenceExpressionComparedZero.ql:36]: Done in 6a97d02247578ee82fc03712918d957eee29ec8b.

**Review by geoffw0 — APPROVED** [review]
LGTM.

File: cpp/change-notes/2020-02-04-unsigned-difference-expression-compared-zero.md (hunk 1 above)
Comment (rdmarsh2) [reply cpp/change-notes/2020-02-04-unsigned-difference-expression-compared-zero.md:1]: That's true by default, but there's alternate suites that are recommended to users that want to run more queries. We might want a separate note for code scanning that mentions the `security-extended` and `security-and-quality` suites.

**Pull request merged**